### ISSUE TITLE
Is circular_queue::safe_reset necessary?

### ISSUE THREAD

**Issue by cinquin** [RichardWarburton/honest-profiler#109]
Clang seems to want to optimize the zeroing in `circular_queue::safe_reset` by a call to `bzero`, even when using `#pragma clang optimize off`. I'm not completely clear as to whether `bzero` is async-safe (this function is defined directly in assembly, at least under FreeBSD, and although it certainly does not call `malloc` I couldn't tell whether clobbering is an issue). I assume that there is probably a problem with `bzero`, given that under FreeBSD and Mac OS X the program appears to get stuck with a number of threads in ``libsystem_platform.dylib`_platform_bzero$VARIANT$Merom``.

I tried just removing the code in `circular_queue::safe_reset`, and nothing terribly bad seemed to happen (I did not check that the stack traces were all correct).

Can `circular_queue::safe_reset` be removed? If not, how can the problem be addressed?

Edit: the use of a volatile variable as below obviously prevents the `bzero` optimization (and allows the program to complete), but it's probably very wasteful:

```
volatile int zero = 0;

void safe_reset(void *start, size_t size) {
    char *base = reinterpret_cast<char *>(start);
    char *end = base + size;
    for (char *p = base; p < end; p++) {
        *p = zero;
    }
}
```

**Comment by RichardWarburton** [comment]
Thanks for investigating. Safe reset, or something like it, has to exist because memset isn't async-safe. The circular buffer design that I'm using here requires zeroing out of things.

GCC was trying to optimise it away to memset before, and we used a specific flag to disable the optimisation. AFAIK bzero is a legacy function and use of memset is preferred.

I think the best solution to this problem would be to find out what the flag is to clang to disable the "optimise reset functions to bzero" optimisation and then apply that.

**Comment by cinquin** [comment]
Right, but I don't know what the right flag would be for Clang.

Would you mind walking me through why zeroing is necessary at all? My analysis is as follows:

- `safe_reset` is only ever called from `Profiler::handle` to zero out `JVMPI_CallFrame frames[kMaxFramesToCapture]`, which is a local variable
- `frames` is used as a field in `trace` passed as an argument to `AsyncGetCallTrace`, which fills in `frames`
- after the call to `AsyncGetCallTrace`, `frames` is passed to `CircularQueue::push`
- `frames` is not read at all in `CircularQueue::push`, but is passed to `CircularQueue::write`
- copies of `lineno` and `method_id` from each element of `frames` are made into `frame_buffer_[slot]`; there is a comment in `CircularQueue::write` that says "Padding already set to 0 by the consumer", but I'm not sure what that means

Unless `AsyncGetCallTrace` needs `frames` to have been pre-zeroed by the caller, as far as I can tell zeroing `frames` has no functional consequence: all the fields in `frames` that are read should have been written to by `AsyncGetCallTrace`, overwriting their previous value. I had a quick look at the Hotspot source and I don't see mention of a requirement for pre-zeroing before calling `AsyncGetCallTrace` (and a cursory examination of the `AsyncGetCallTrace` code also suggests it does not rely on pre-zeroing, as far as I can tell).

I'm not familiar with the history of why things are set up in this way, so please let me know if I'm missing something.

**Comment by RichardWarburton** [comment]
Sure, finding the correct flag will need a bit of research if we go down that route.

For further explanation of the padding comment, see issue #77. For more information on struct padding see this SO question - http://stackoverflow.com/questions/5397447/struct-padding-in-c.

The safe_reset function's declaration actually includes a comment that explains the zeroing out issue and has been in the codebase ever since I took Jeremy Manson's original prototype.

```
// We have to set every byte to 0 instead of just initializing the
// individual fields, because the structs might be padded, and we
// use memcmp on it later.  We can't use memset, because it isn't
// async-safe.
```

Now an interesting observation is that we actually don't memcpy the structs anymore and haven't for an absolute age. So we can possibly just initialize the fields. We do need to initialize the fields though because otherwise we're passing a pointer to uninitialized data to a function. If we do make that change though - we'll /still/ have to check that clang and GCC don't try to optimise/breakify everything to memset or bzero.

**Comment by cinquin** [comment]
Overall my point is that `AsyncGetCallTrace` does not seem to need its `JVMPI_CallTrace` argument zeroed out. It seems to me that it's fine to pass  to a function a pointer to uninitialized memory, as long as the function only ever writes to that memory.

If one wants to make extra sure that no uninitialized data are written out because of struct padding, it does make sense to zero out records in the queue. [This comment](https://github.com/RichardWarburton/honest-profiler/issues/77#issuecomment-69375322) in issue #77 suggested having the zeroing be done by the consumer instead of the producer, which I think makes good sense.

To summarize, removing the zeroing in `Profiler::handle` should not affect the output of the profiler, but if anything it should improve performance (which is especially welcome in the signal handler context), and it would do away with the async safety problem.

**Comment by nitsanw** [comment]
@cinquin I think you are correct and that safe_reset is not required.

**Comment by nitsanw** [comment]
Also, given that we currently ALWAYS publish a trace, we can avoid copying all together and use the queue allocated call trace. This will end up being a claim/release kind of interface rather than the current 'push'.

**Comment by cinquin** [comment]
@nitsanw I agree that a claim/release interface would be nice since it would minimize unnecessary copying. It would require changes in the `CircularQueue` code though. To avoid making things more complex, for now I submitted pull request #110 that just removes `safe_reset`. I did not touch the GCC compile option in case it has effects in other places.

**Comment by RichardWarburton** [comment]
Thanks for the discussion gents.

re: "This comment in issue #77 suggested having the zeroing be done by the consumer instead of the producer, which I think makes good sense." - it actually already does, that refers to zeroing out the data structure in the queue itself though.

I'm happy with @nitsanw's suggestion of just passing a pointer to the queue directly with a claim/release API, makes sense from an efficiency point of view.

I'll take a look into the uninitialized data situation. I was under the impression that passing uninitialized data to a function had a boatload of undefined behaviour. If that's not correct then I'm happy to accept #110, otherwise I think the pointer still need initializing.

**Comment by RichardWarburton** [comment]
I've accepted the pull request.

**Comment by cinquin** [comment]
Great; thanks.
I've updated my other pull request #108  per your comments. With these two pull requests, honest profiler should work pretty well both under Mac OS X and FreeBSD.
